zephyr/lib/vpages.c

```diff
@@ -0,0 +1,340 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright(c) 2025 Intel Corporation.
+ *
+ * Author: Liam Girdwood <liam.r.girdwood@linux.intel.com>
+ */
+
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <sof/lib/regions_mm.h>
+#include <zephyr/drivers/mm/mm_drv_intel_adsp_mtl_tlb.h>
+
+LOG_MODULE_REGISTER(vpage, CONFIG_SOF_LOG_LEVEL);
+
+/* Simple Page Allocator.
+ *
+ * This allocator manages the allocation and deallocation of virtual memory pages from
+ * a predefined virtual memory region which is larger than the physical memory region.
+ *
+ * Both memory regions are divided into 4kB pages that are represented as blocks in a
+ * bitmap using the zephyr sys_mem_blocks API. The virtual block map tracks the allocation
+ * of virtual memory pages while the physical block map in the Zephyr MM driver tracks
+ * the allocation of physical memory pages.
+ */
+
+/* max number of allocation elements */
+#define VPAGE_MAX_ALLOCS	CONFIG_SOF_VPAGE_MAX_ALLOCS
+
+/*
+ * Virtual memory allocation element - tracks allocated virtual page id and size
+ */
+struct valloc_elem {
+	uint16_t pages;		/* number of 4kB pages allocated in contiguous block */
+	uint16_t vpage;		/* virtual page number from start of region */
+} __packed;
+
+/*
+ * Virtual page table structure
+ *
+ * This structure holds all information about virtual memory pages
+ * including the number of free and total pages, the virtual memory
+ * region, the block allocator for virtual pages and the allocation
+ * elements.
+ */
+struct vpage_context {
+	struct k_mutex lock;
+	uint32_t free_pages;	/* number of free 4kB pages */
+	uint32_t total_pages;	/* total number of 4kB pages */
+
+	/* Virtual memory region information */
+	const struct sys_mm_drv_region *virtual_region;
+	struct sys_mem_blocks vpage_blocks;
+
+	/* allocation elements to track page id to allocation size */
+	uint32_t num_elems_in_use;	/* number of allocated elements in use*/
+	struct valloc_elem velems[VPAGE_MAX_ALLOCS];
```

Comment:
`VPAGE_MAX_ALLOCS` is per allocator. So if we have several modules, each using this allocator, each will have up to that many allocations. But this structure is global and only present once in the system?